src/marten/db/query/related_set.cr

```diff
@@ -15,6 +15,11 @@ module Marten
                    end
         end
 
+        protected def fetch
+          super
+          @result_cache.not_nil!.each { |r| r.assign_related_object(@instance, @related_field_id) }
```

Comment:
I don't think that this should be the default behaviour for this query set abstraction. The issue I see here is that we are making the assumption that the `Marten::DB::Query::RelatedSet` abstraction is (or will always be) used for reverse relations only and that setting this object automatically will always make sense, but that may not necessarily be the case! A library could well make use of this abstraction for creating query sets in different contexts, and preloading associated objects might not necessarily be wanted in these situations.

My recommendation would be to:

1. Add a new `assign_related` bool argument to the `#initialize` method of this abstraction that would default to `false`.
2. Make sure that `assign_related: true` is specified when creating the query sets corresponding to reverse many-to-one relations [here](https://github.com/martenframework/marten/blob/c28cd38c38f9ff941c803768051af6a31c21914e/src/marten/db/field/many_to_one.cr#L205).
3. Make sure that the related object instance is assigned only when `assign_related` is set to `true`.